File: clients/go/frame/frame.go
// Package frame implements the native Volant 16-byte frame header.
//
// On-wire layout is big-endian (matches crates/volant-protocol/src/codec.rs):
//
//	magic u8 | version u8 | opcode u16 | correlation_id u32 | payload_len u32 | crc32 u32 | payload
//
// CRC32 is IEEE of the payload only (hash/crc32 ≡ zlib.crc32 ≡ crc32fast).
package frame

import (
	"encoding/binary"
	"fmt"
	"hash/crc32"
)

const (
	// Magic is the frame identifier byte 'V'.
	Magic = 0x56
	// ProtocolVersion is the only accepted header version.
	ProtocolVersion = 1
	// HeaderLen is the on-wire header size in bytes.
	HeaderLen = 16
	// MaxPayload is the 16 MiB payload cap.
	MaxPayload = 16 * 1024 * 1024
)

// ProtocolError is a magic, version, checksum, or framing error.
type ProtocolError struct {
	Msg string
}

func (e *ProtocolError) Error() string { return e.Msg }

// Frame is a decoded native-protocol frame.
type Frame struct {
	Opcode        uint16
	CorrelationID uint32
	Payload       []byte
	Version       uint8
	Checksum      uint32
}

// Checksum returns the IEEE CRC32 of payload (same polynomial as crc32fast).
func Checksum(payload []byte) uint32 {
	return crc32.ChecksumIEEE(payload)
}

// Encode writes a complete frame (header + payload) at protocol version 1.
func Encode(opcode uint16, correlationID uint32, payload []byte) ([]byte, error) {
	return EncodeVersion(opcode, correlationID, payload, ProtocolVersion)
}

// EncodeVersion writes a frame with an explicit version (tests use this to
// produce a rejected version byte).
func EncodeVersion(opcode uint16, correlationID uint32, payload []byte, version uint8) ([]byte, error) {
	if len(payload) > MaxPayload {
		return nil, &ProtocolError{Msg: fmt.Sprintf("payload too large: %d > %d", len(payload), MaxPayload)}
	}
	crc := Checksum(payload)
	out := make([]byte, HeaderLen+len(payload))
	out[0] = Magic
	out[1] = version
	binary.BigEndian.PutUint16(out[2:4], opcode)
	binary.BigEndian.PutUint32(out[4:8], correlationID)
	binary.BigEndian.PutUint32(out[8:12], uint32(len(payload)))
	binary.BigEndian.PutUint32(out[12:16], crc)
	copy(out[HeaderLen:], payload)
	return out, nil
}

// TryDecode parses one frame from data.
//
// Returns (nil, data, nil) if more bytes are needed.
// Returns a non-nil error on magic / version / checksum mismatch.
func TryDecode(data []byte) (*Frame, []byte, error) {
	if len(data) < HeaderLen {
		return nil, data, nil
	}
	magic := data[0]
	version := data[1]
	opcode := binary.BigEndian.Uint16(data[2:4])
	corr := binary.BigEndian.Uint32(data[4:8])
	payloadLen := binary.BigEndian.Uint32(data[8:12])
	crcWire := binary.BigEndian.Uint32(data[12:16])
	if magic != Magic {
		return nil, data, &ProtocolError{Msg: fmt.Sprintf("invalid frame magic: %#x", magic)}
	}
	if payloadLen > MaxPayload {
		return nil, data, &ProtocolError{Msg: fmt.Sprintf("payload too large: %d > %d", payloadLen, MaxPayload)}
	}
	total := HeaderLen + int(payloadLen)
	if len(data) < total {
		return nil, data, nil
	}
	payload := make([]byte, payloadLen)
	copy(payload, data[HeaderLen:total])
	if version != ProtocolVersion {
		return nil, data, &ProtocolError{Msg: fmt.Sprintf("unsupported protocol version: %d", version)}
	}
	expected := Checksum(payload)
	if crcWire != expected {
		return nil, data, &ProtocolError{Msg: fmt.Sprintf("checksum mismatch: got %#x, expected %#x", crcWire, expected)}
	}
	return &Frame{
		Opcode:        opcode,
		CorrelationID: corr,
		Payload:       payload,
		Version:       version,
		Checksum:      crcWire,
	}, data[total:], nil
}

// Decode parses a complete frame. It errors if the buffer is truncated,
// invalid, or has trailing bytes after one frame.
func Decode(data []byte) (*Frame, error) {
	f, rest, err := TryDecode(data)
	if err != nil {
		return nil, err
	}
	if f == nil {
		return nil, &ProtocolError{Msg: "incomplete frame"}
	}
	if len(rest) > 0 {
		return nil, &ProtocolError{Msg: fmt.Sprintf("trailing bytes after frame: %d", len(rest))}
	}
	return f, nil
}

File: clients/python/src/volant/frame.py
"""Native Volant frame encode/decode.

On-wire header is 16 bytes, big-endian (matches `bytes::BufMut::put_u16` /
`put_u32` in `crates/volant-protocol/src/codec.rs`):

    magic u8 | version u8 | opcode u16 | correlation_id u32 | payload_len u32 | crc32 u32 | payload

CRC32 is IEEE of the **payload only** (`crc32fast` / `zlib.crc32`).
"""

from __future__ import annotations

import struct
import zlib
from dataclasses import dataclass

FRAME_MAGIC = 0x56  # b'V'
PROTOCOL_VERSION = 1
HEADER_LEN = 16
MAX_PAYLOAD = 16 * 1024 * 1024

_HEADER_STRUCT = struct.Struct(">BBHIII")


class ProtocolError(Exception):
    """Magic, version, checksum, or framing error."""


@dataclass(frozen=True)
class Frame:
    opcode: int
    correlation_id: int
    payload: bytes
    version: int = PROTOCOL_VERSION
    checksum: int = 0


def checksum(payload: bytes) -> int:
    """IEEE CRC32 of `payload` (same polynomial as Rust `crc32fast::hash`)."""
    return zlib.crc32(payload) & 0xFFFFFFFF


def encode_frame(
    opcode: int,
    correlation_id: int,
    payload: bytes,
    *,
    version: int = PROTOCOL_VERSION,
) -> bytes:
    """Encode a complete frame (header + payload)."""
    if len(payload) > MAX_PAYLOAD:
        raise ProtocolError(f"payload too large: {len(payload)} > {MAX_PAYLOAD}")
    crc = checksum(payload)
    header = _HEADER_STRUCT.pack(
        FRAME_MAGIC, version, opcode, correlation_id, len(payload), crc
    )
    return header + payload


def try_decode_frame(data: bytes) -> tuple[Frame | None, bytes]:
    """Decode one frame from `data`.

    Returns `(None, data)` if more bytes are needed.
    Raises :class:`ProtocolError` on magic / version / checksum mismatch.
    """
    if len(data) < HEADER_LEN:
        return None, data
    magic, version, opcode, corr, payload_len, crc_wire = _HEADER_STRUCT.unpack_from(
        data, 0
    )
    if magic != FRAME_MAGIC:
        raise ProtocolError(f"invalid frame magic: {magic:#x}")
    if payload_len > MAX_PAYLOAD:
        raise ProtocolError(f"payload too large: {payload_len} > {MAX_PAYLOAD}")
    total = HEADER_LEN + payload_len
    if len(data) < total:
        return None, data
    payload = data[HEADER_LEN:total]
    if version != PROTOCOL_VERSION:
        raise ProtocolError(f"unsupported protocol version: {version}")
    expected = checksum(payload)
    if crc_wire != expected:
        raise ProtocolError(
            f"checksum mismatch: got {crc_wire:#x}, expected {expected:#x}"
        )
    frame = Frame(
        opcode=opcode,
        correlation_id=corr,
        payload=payload,
        version=version,
        checksum=crc_wire,
    )
    return frame, data[total:]


def decode_frame(data: bytes) -> Frame:
    """Decode a complete frame. Raises if truncated or invalid."""
    frame, rest = try_decode_frame(data)
    if frame is None:
        raise ProtocolError("incomplete frame")
    if rest:
        raise ProtocolError(f"trailing bytes after frame: {len(rest)}")
    return frame
